akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -738,38 +741,56 @@ private[akka] class ShardRegion(
   }
 
   def replyToRegionStateQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.CurrentShardState](Shard.GetCurrentShardState)
-      .map { shardStates =>
-        CurrentShardRegionState(shardStates.map {
-          case (shardId, state) => ShardRegion.ShardState(shardId, state.entityIds)
-        }.toSet)
-      }
-      .recover {
-        case _: AskTimeoutException => CurrentShardRegionState(Set.empty)
+    queryShards[Shard.CurrentShardState](shards, Shard.GetCurrentShardState)
+      .map { qr =>
+        // TODO add CurrentShardRegionState field for timeouts (proto). Until then
+        val state = qr.responses.map(state => ShardRegion.ShardState(state.shardId, state.entityIds)) ++
+          qr.timedout.map(ShardRegion.ShardState(_, Set.empty))
+        CurrentShardRegionState(state.toSet)
       }
       .pipeTo(ref)
   }
 
   def replyToRegionStatsQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.ShardStats](Shard.GetShardStats)
-      .map { shardStats =>
-        ShardRegionStats(shardStats.map {
-          case (shardId, stats) => (shardId, stats.entityCount)
-        }.toMap)
-      }
-      .recover {
-        case _: AskTimeoutException => ShardRegionStats(Map.empty)
+    queryShards[ShardStats](shards, Shard.GetShardStats)
+      .map { qr =>
+        // TODO add ShardRegionStats field for timeouts (proto): ShardRegion.ShardRegionStats(stats, timeouts.tosSet)
+        // Until then
+        ShardRegionStats(qr.responses.map(stats => (stats.shardId, stats.entityCount)).toMap ++ qr.timedout.map((_, 0)))
       }
       .pipeTo(ref)
   }
 
-  def askAllShards[T: ClassTag](msg: Any): Future[Seq[(ShardId, T)]] = {
+  /**
+   * Query all or a subset of shards, e.g. unresponsive shards that initially timed out.
+   * If the number of `shards` are less than this.shards.size, this could be a retry.
+   * Returns a partitioned set of any shards that may have not replied within the
+   * timeout and shards did reply, to provide retry on only that subset.
+   *
+   * Logs a warning if any of the group timed out.
+   *
+   * To check subset unresponsive: {{{ queryShards[T](shards.filterKeys(u.contains), shardQuery) }}}
+   */
+  def queryShards[T: ClassTag](shards: Map[ShardId, ActorRef], msg: Any): Future[ShardsQueryResult[ShardId, T]] = {
     implicit val timeout: Timeout = settings.shardRegionQueryTimeout
-    Future.sequence(shards.toSeq.map {
-      case (shardId, ref) => (ref ? msg).mapTo[T].map(t => (shardId, t))
-    })
+    Future
+      .sequence(shards.toSeq.map {
+        case (shardId, shard) => askOne(shard, msg, shardId)
+      })
+      .map { ps =>
+        val qr = ShardsQueryResult[ShardId, T](ps, shards.size)
```

Comment:
It would be the same in this PR but in future the idea is it allows us internally retry a subset of those that timed out or other use case for a subset versus all. This is the thinking behind `queryShards(n)` vs allShards. `queryShards` doesn't use `this.shards`.